experimental/oak_baremetal_channel/src/frame.rs

```diff
@@ -0,0 +1,161 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+extern crate alloc;
+
+use alloc::{format, vec, vec::Vec};
+use bitflags::bitflags;
+use ciborium_io::{Read, Write};
+
+pub const PADDING_SIZE: usize = 4;
+
+type Length = u16;
+pub const LENGTH_SIZE: usize = 2;
+static_assertions::assert_eq_size!([u8; LENGTH_SIZE], Length);
+
+bitflags! {
+    #[derive(Default)]
+    pub struct Flags: u16 {
+        const START = 1;
+        const END = 2;
+    }
+}
+pub const FLAGS_SIZE: usize = 2;
+static_assertions::assert_eq_size!([u8; FLAGS_SIZE], Flags);
+
+pub const BODY_OFFSET: usize = 8;
+static_assertions::assert_eq_size!(
+    [u8; BODY_OFFSET],
+    [u8; PADDING_SIZE + LENGTH_SIZE + FLAGS_SIZE]
+);
+
+pub const MAX_SIZE: usize = 64000;
+pub const MAX_BODY_SIZE: usize = MAX_SIZE - BODY_OFFSET;
+
+#[derive(Clone, Default, Debug)]
+pub struct Frame {
+    pub flags: Flags,
+    pub body: Vec<u8>,
+}
+
+impl Frame {
+    fn len(&self) -> anyhow::Result<Length> {
+        let length = BODY_OFFSET + self.body.len();
+
+        if length > MAX_SIZE {
+            return Err(anyhow::Error::msg("frame exceeds the maximum frame size"));
+        };
+
+        Length::try_from(length).map_err(|_error| {
+            anyhow::Error::msg(format!(
+                "could convert the frame length usize to {:?}",
+                core::any::type_name::<Length>()
+            ))
+        })
```

Comment:
This seems like it is trying to second guess the lenght of the serialized message, but it is actually quite coupled with the serialization implementation. In fact, the serialization method seems to rely on this method, which seems backwards. My preference would be to inline this calculation in the serialization method, and not expose a len method at all; or, if we want to expose it, it should rely on the actual serialized format.
I don't feel too strongly about this, but at least some comments would help to make sure that this is well understood.